src/neorados/cls/rgw.h

```diff
@@ -0,0 +1,295 @@
+[[nodiscard]] inline
+auto bi_clean()
+{
+  return ClsWriteOp{[] (WriteOp& op) {
+    op.remove();
+  }};
+}
+
+/// \brief Set the resharding status of a bucket index shard object.
+///
+/// \return The ClsWriteOp to be passed to WriteOp::exec
+[[nodiscard]] inline
+auto bi_set_resharding(cls_rgw_reshard_status status)
+{
+  const auto call = cls_rgw_set_bucket_resharding_op{
+    .entry = {.reshard_status = status}
+  };
+  buffer::list in;
+  encode(call, in);
+  return ClsWriteOp{[in = std::move(in)] (WriteOp& op) {
+    op.exec(RGW_CLASS, RGW_SET_BUCKET_RESHARDING, in);
+  }};
+}
+
+/// \brief Write an entry directly to the bucket index shard object.
+///
+/// \return The ClsWriteOp to be passed to WriteOp::exec
+[[nodiscard]] inline
+auto bi_put(rgw_cls_bi_entry entry)
+{
+  const auto call = rgw_cls_bi_put_op{.entry = std::move(entry)};
+  buffer::list in;
+  encode(call, in);
+  return ClsWriteOp{[in = std::move(in)] (WriteOp& op) {
+    op.exec(RGW_CLASS, RGW_BI_PUT, in);
+  }};
+}
+
+/// An input iterator whose value type is convertible to rgw_cls_bi_entry.
+template <typename T>
+concept bi_entry_iterator =
+    std::input_iterator<T> &&
+    std::convertible_to<std::iter_value_t<T>, rgw_cls_bi_entry>;
+
+/// An input range whose value type is convertible to rgw_cls_bi_entry.
+template <typename T>
+concept bi_entry_range =
+    std::ranges::input_range<T> &&
+    bi_entry_iterator<std::ranges::iterator_t<T>>;
+
+/// \brief Write entries directly to the bucket index shard object.
+///
+/// \param begin Beginning of a range of entries.
+/// \param end End of a range of entries.
+/// \param check_existing Subtract bucket stats for overwritten entries.
+///
+/// \return The ClsWriteOp to be passed to WriteOp::exec
+[[nodiscard]] inline
+auto bi_put_entries(bi_entry_iterator auto begin,
+                    bi_entry_iterator auto end,
+                    bool check_existing)
+{
+  const auto call = rgw_cls_bi_put_entries_op{
+    .entries = {begin, end},
+    .check_existing = check_existing
+  };
+  bufferlist in;
+  encode(call, in);
+  return ClsWriteOp{[in = std::move(in)] (WriteOp& op) {
+    op.exec(RGW_CLASS, RGW_BI_PUT_ENTRIES, in);
+  }};
+}
+
+/// \overload
+[[nodiscard]] inline
+auto bi_put_entries(bi_entry_range auto stats)
+{
+  // forward to iterator overload
+  return bi_put_entries(std::ranges::begin(stats),
+                        std::ranges::end(stats));
+}
+
+/// \brief List bucket index entries.
+///
+/// \param name_filter List entries whose key starts with the given prefix.
+/// \param marker Resume listing after the given marker position.
+/// \param max Maximum number of entries to return.
+/// \param reshardlog When true, list entries from the reshard log.
+/// \param out Output iterator to receive the entries.
+/// \param truncated Output parameter set to true if more entries follow.
+///
+/// \return The ClsReadOp to be passed to ReadOp::exec
+[[nodiscard]] inline
+auto bi_list(std::string name_filter,
```

Comment:
I've been using span-in/span-out in other parts of CLS and I've found it rather unwieldy and thinking of similarly just switching everything over to use an output iterator instead.

Reply:
still on the fence about this part. the iterator/range concepts are great for generic algorithms, but may be overly general for the cls interface

in this case, it'd be better to return entries as `std::vector<rgw_cls_bi_entry>&` so we can move-assign from the decoded `rgw_cls_bi_list_ret::entries` vector

my gripe with the existing cls_rgw_client.h interface isn't necessarily that it bakes in specific container types, but that it chose inefficient ones like std::list and std::map

Reply:
That's reasonable since we own both sides of it. If we don't like the data structure it's using we can just change it.